### PR TITLE
compiler: Allow mutating refs in callbacks passed to JSX

### PR DESCRIPTION
**Pull request by josephsavona**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #29733

Fixes false positives where we currently disallow mutations of refs from callbacks passed to JSX, if the ref is also passed to jsx. We consider these to be mutations of "frozen" values, but refs are explicitly allowed to have interior mutability. The fix is to always allow (at leat within InferReferenceEffects) for refs to be mutated. This means we completely rely on ValidateNoRefAccessInRender to validate ref access and stop reporting false positives.

### CODE REVIEW

**Comment by josephsavona** [comment]
Fixes #29741 
Fixes #29703

File: compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts [hunk 1]
```diff
@@ -521,7 +525,28 @@ class InferenceState {
         break;
       }
       case Effect.Mutate: {
-        if (valueKind.kind === ValueKind.Context) {
+        if (
+          isRefValueType(place.identifier) ||
+          isUseRefType(place.identifier)
+        ) {
+          if (this.#env.config.validateRefAccessDuringRender) {
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts:532]: Previously refs would have fallen through to the  case where the valueKind is mutable, and we would record it as a mutation. Now we always treat refs specially, and either ignore them (default) or record a RefMutation error when ref validation is on.

Ref validation still has a bunch of false positives that we have to fix before we can enable it more broadly. However, the fix here avoids the false negatives even when the validation is off.

**Review by gsathya — APPROVED** [review]
awesome!

**Comment by josephsavona** [comment]
Heads up that i updated this to remove the RefMutation variant of FunctionEffect, we can keep this simple by just ignoring ref mutations completely in InferReferenceEffects.

**Pull request merged**